crates/driver/src/boundary/settlement.rs

```diff
@@ -245,6 +248,115 @@ impl Settlement {
             solver: self.solver,
         })
     }
+
+    pub fn with_protocol_fees(&mut self, solution: &competition::Solution) {
+        struct PricedTrade {
+            sell_price: eth::U256,
+            buy_price: eth::U256,
+        }
+
+        let boundary_trades = self
+            .inner
+            .encoder
+            .user_trades()
+            .map(|trade| {
+                (
+                    trade.data.order.metadata.uid.0.into(),
+                    PricedTrade {
+                        sell_price: trade.sell_token_price,
+                        buy_price: trade.buy_token_price,
+                    },
+                )
+            })
+            .collect::<HashMap<Uid, PricedTrade>>();
+
+        let new_boundary_trades = solution
+            .trades()
+            .iter()
+            .filter_map(|trade| {
+                match trade {
```

Comment:
Can this logic be implemented on an `impl Trade` extension in this file?

Maybe we can have multiple methods instead of one giant 5 indentations deep block, with descriptive names of each step of the adjustment process:
1. Get expected amounts
2. Get executed amount
3. Compute difference between 1 and 2 and applying fee policy to it
4. Adjust executed amounts on the trade down by the right difference